compute_tools/src/bin/compute_ctl.rs

```diff
@@ -363,7 +364,15 @@ fn main() -> Result<()> {
             .map_err(|e| tracing::error!("log thread panicked: {:?}", e));
 
         info!("Postgres exited with code {}, shutting down", ecode);
-        exit_code = ecode.code()
+        exit_code = ecode.code();
+        let mut state = compute.state.lock().unwrap();
+        if state.status == ComputeStatus::TerminationPending {
+            // we were asked to terminate gracefully, don't exit to avoid restart
+            delay_exit = true
+        }
+        state.status = ComputeStatus::Terminated;
+        compute.state_changed.notify_all();
+        drop(state);
```

Comment:
Bellow there is a code for syncing safekeepers at the compute exit. So if cplane waits for `terminated` status, then this opens a small window for a race:
1. we marked compute as terminated
2. cplane marked endpoint as idle and compute as terminated
3. sync safekeepers is still running
4. cplane can decide to start compute again

In this script, if 3. will take longer for some reason, but 4. is quick enough, we can have a case when two walproposers are running at the same time. IIRC, safekeepers are specifically designed to protect from this, but this is just not fully correct and they will kick each other

I think it could be solved by simply moving this new code block with Terminated state setting right after optional sync safekeepers. Then cplane will only move forward when there is no activity inside compute node

Reply:
Sounds reasonable, thanks!